File: training-data/typescript/happy_number.ts
const digitSquareSum = (n: number): number => [...String(n)].reduce((acc, d) => acc + Number(d) ** 2, 0);

export function isHappy(n: number): boolean {
  const seen = new Set<number>();
  while (n !== 1 && !seen.has(n)) {
    seen.add(n);
    n = digitSquareSum(n);
  }
  return n === 1;
}

console.log(Array.from({ length: 49 }, (_, i) => i + 1).filter(isHappy));

File: training-data/rust/happy_number.rs
use std::collections::HashSet;

fn digit_square_sum(mut n: u32) -> u32 {
    let mut s = 0;
    while n > 0 {
        s += (n % 10).pow(2);
        n /= 10;
    }
    s
}

fn is_happy(mut n: u32) -> bool {
    let mut seen = HashSet::new();
    while n != 1 && seen.insert(n) {
        n = digit_square_sum(n);
    }
    n == 1
}

fn main() {
    let happy: Vec<u32> = (1..=50).filter(|&n| is_happy(n)).collect();
    println!("{:?}", happy);
    println!("count up to 10000: {}", (1..=10000).filter(|&n| is_happy(n)).count());
}
